polkadot/consensus/src/lib.rs

```diff
@@ -495,133 +252,211 @@ impl<C: PolkadotApi, N: Network> bft::ProposerFactory for ProposerFactory<C, N>
-					continue
-				}
+		let current_timestamp = current_timestamp();
 
-				if pending_size + pending.encoded_size() >= MAX_TRANSACTIONS_SIZE { break }
+		// do initial serialization and structural integrity checks.
+		let maybe_proposal = evaluation::evaluate_initial(
+			proposal,
+			current_timestamp,
+			&self.parent_hash,
+			self.parent_number,
+			&active_parachains,
+		);
 
-				match block_builder.push_extrinsic(pending.as_transaction().clone()) {
-					Ok(()) => {
-						pending_size += pending.encoded_size();
-					}
-					Err(_) => {
-						unqueue_invalid.push(pending.hash().clone());
-					}
-				}
+		let proposal = match maybe_proposal {
+			Ok(p) => p,
+			Err(e) => {
+				// TODO: these errors are easily re-checked in runtime.
+				debug!(target: "bft", "Invalid proposal: {:?}", e);
+				return Box::new(future::ok(false));
 			}
+		};
 
-			for tx_hash in unqueue_invalid {
-				pool.remove(&tx_hash, false);
-			}
-		}
+		let vote_delays = {
+			// delay casting vote until able (according to minimum block time)
+			let minimum_delay = self.delay.clone()
+				.map_err(|e| Error::from(::tokio_timer::TimerError::clone(&*e)));
+
+			let included_candidate_hashes = proposal
+				.parachain_heads()
+				.iter()
+				.map(|candidate| candidate.hash());
+
+			// delay casting vote until we have proof that all candidates are
+			// includable.
+			let includability_tracker = self.table.track_includability(included_candidate_hashes)
+				.map_err(|_| ErrorKind::PrematureDestruction.into());
+
+			// the duration at which the given number of parachains is acceptable.
+			let count_delay = self.dynamic_inclusion.acceptable_in(
+				Instant::now(),
+				proposal.parachain_heads().len(),
+			);
+
+			// the duration until the given timestamp is current
+			let proposed_timestamp = proposal.timestamp();
+			let timestamp_delay = if proposed_timestamp > current_timestamp {
+				Some(Duration::from_secs(proposed_timestamp - current_timestamp))
+			} else {
+				None
+			};
 
-		let polkadot_block = block_builder.bake();
-		info!("Proposing block [number: {}; hash: {}; parent_hash: {}; extrinsics: [{}]]",
-			polkadot_block.header.number,
-			Hash::from(polkadot_block.header.blake2_256()),
-			polkadot_block.header.parent_hash,
-			polkadot_block.extrinsics.iter()
-				.map(|xt| format!("{}", Hash::from(xt.blake2_256())))
-				.collect::<Vec<_>>()
-				.join(", ")
-		);
+			// construct a future from the maximum of the two durations.
+			let temporary_delay = match ::std::cmp::max(timestamp_delay, count_delay) {
+				Some(duration) => {
+					let f = self.timer.sleep(duration).map_err(Error::from);
+					future::Either::A(f)
+				}
+				None => future::Either::B(future::ok(())),
+			};
 
-		let substrate_block = Slicable::decode(&mut polkadot_block.encode().as_slice())
-			.expect("polkadot blocks defined to serialize to substrate blocks correctly; qed");
+			minimum_delay.join3(includability_tracker, temporary_delay)
```

Comment:
evaluation of proposed blocks happens immediately, but we don't cast a "good" vote until 
  1. the timestamp is valid
  2. the number of included parachain candidates falls below the inclusion curve
  3. each included parachain candidate has the requisite number of votes